jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java

```diff
@@ -281,23 +278,39 @@ private static final Object resolveInvokeDynamic(long j9class, String name, Stri
 
 			/* result[1] stores a MethodHandle object, which is used as the last argument to the caller method. */
 			result[1] = appendixResult[0];
-/*[IF JAVA_SPEC_VERSION < 11]*/
 		} catch (Throwable e) {
-			if (type == null) {
-				throw new BootstrapMethodError(e);
-			}
-
-			/* linkCallSite may correctly throw a BootstrapMethodError. */
+/*[IF JAVA_SPEC_VERSION < 11]*/
+			/* Before Java 11, if linkCallSite throws a BootstrapMethodError due to the CallSite binding
+			 * resolving to null, throw it at resolution time. This ensures that resolution will be
+			 * reattempted on the subsequent visit to the invokedynamic instruction in the interpreter.
+			 */
 			if (e instanceof BootstrapMethodError) {
 				throw e;
 			}
-
-			/* Any other throwables are wrapped in an invoke-time BootstrapMethodError exception throw. */
+/*[ENDIF] JAVA_SPEC_VERSION < 11*/
+			/* Any throwables are wrapped in an invoke-time BootstrapMethodError exception throw. */
 			try {
+				MethodHandle resultHandle;
 				MethodHandle thrower = MethodHandles.throwException(type.returnType(), BootstrapMethodError.class);
 				MethodHandle constructor = IMPL_LOOKUP.findConstructor(BootstrapMethodError.class, MethodType.methodType(void.class, Throwable.class));
-				MethodHandle resultHandle = MethodHandles.foldArguments(thrower, constructor.bindTo(e));
+
+				MethodHandle combiner = null;
+				if (e instanceof BootstrapMethodError) {
+					Throwable cause = e.getCause();
+					if (cause == null) {
+						combiner = constructor.bindTo(e.getMessage());
+					} else {
+						combiner = constructor.bindTo(cause);
+					}
+				} else {
+					combiner = constructor.bindTo(e);
+				}
+				resultHandle = MethodHandles.foldArguments(thrower, combiner);
+/*[IF JAVA_SPEC_VERSION >= 11]*/
+				MemberName memberName = resultHandle.internalForm().vmentry;
+/*[ELSE] JAVA_SPEC_VERSION >= 11*/
 				MemberName memberName = resultHandle.internalForm().compileToBytecode();
+/*[ENDIF] JAVA_SPEC_VERSION >= 11*/
 				result[0] = memberName;
 				result[1] = resultHandle;
```

Comment:
a potential improvement to get rid of local variables (`resultHandle` and `memberName`) if we modify this file in the future:
```
	result[1] = MethodHandles.foldArguments(thrower, combiner);
/*[IF JAVA_SPEC_VERSION >= 11]*/
	result[0] = result[1].internalForm().vmentry;
/*[ELSE] JAVA_SPEC_VERSION >= 11*/
	result[0] = result[1].internalForm().compileToBytecode();
/*[ENDIF] JAVA_SPEC_VERSION >= 11*/
```